daliuge-engine/dlg/drop.py

```diff
@@ -164,6 +164,32 @@ class AbstractDROP(EventFirer, EventHandler):
     #  - Subclasses implement methods decorated with @abstractmethod
     __metaclass__ = ABCMeta
 
+    # Matcher used to validate environment_variable_syntax
+    _env_var_matcher = re.compile(r"\$[A-z|\d]+\..+")
+    _dlg_var_matcher = re.compile(r"\$DLG_.+")
+
+    _known_locks = ('_finishedProducersLock', '_refLock')
+    _known_rlocks = ('_statusLock',)
+
+    def __getstate__(self):
+        state = self.__dict__.copy()
```

Comment:
any particular reason why you're not using `super().__getstate__()`? Same for `__setstate__`.

Reply:
That's a good question for your past self :stuck_out_tongue: Though since I altered your original commits and `git blame` has my name next it so I suppose I can take the fall :wink: Trying this gives errors with `AttributeError: 'super' object has no attribute '__getstate__'.`, so I guess for some reason the base classes don't provide that? `EventHandler` is an `ABC` so I thought that might be why, but even `super(EventFirer).__getstate__()` doesn't work. Though given this is the base `AbstractDROP` perhaps we want to retain control in here anyway?

Reply:
I _knew_ I would end up fighting with myself! OK, an error is an error. And now thinking about it, `object` ultimately doesn't implement `__{g,s}etstate__`, so it makes sense we do our own thing here. Sorry for the noise!